src/Middleware/RequestDecompression/src/DeflateDecompressionProvider.cs

```diff
@@ -13,6 +14,221 @@ internal sealed class DeflateDecompressionProvider : IDecompressionProvider
+            }
+
+            public override bool CanRead => true;
+            public override bool CanWrite => false;
+            public override bool CanSeek => _stream.CanSeek;
+            public override long Length => _stream.Length;
+            public override long Position { get => _stream.Position; set => _stream.Position = value; }
+            public override long Seek(long offset, SeekOrigin origin) => _stream.Seek(offset, origin);
+            public override void SetLength(long value) => _stream.SetLength(value);
+            public override void Flush() => _stream.Flush();
+
+            public override void Write(byte[] buffer, int offset, int count) => throw new NotSupportedException();
+            public override ValueTask WriteAsync(ReadOnlyMemory<byte> buffer, CancellationToken cancellationToken) => throw new NotSupportedException();
+
+            public int PeekFirstByte()
+            {
+                Debug.Assert(_firstByteStatus == FirstByteStatus.None);
+
+                int value = _stream.ReadByte();
+                if (value == -1)
+                {
+                    _firstByteStatus = FirstByteStatus.Consumed;
+                    return -1;
+                }
+
+                _firstByte = (byte)value;
+                _firstByteStatus = FirstByteStatus.Available;
+                return value;
+            }
+
+            public async ValueTask<int> PeekFirstByteAsync(CancellationToken cancellationToken)
+            {
+                Debug.Assert(_firstByteStatus == FirstByteStatus.None);
+
+                var buffer = new byte[1];
+
+                int bytesRead = await _stream.ReadAsync(buffer, cancellationToken).ConfigureAwait(false);
+                if (bytesRead == 0)
+                {
+                    _firstByteStatus = FirstByteStatus.Consumed;
+                    return -1;
+                }
+
+                _firstByte = buffer[0];
+                _firstByteStatus = FirstByteStatus.Available;
+                return buffer[0];
+            }
+
+            public override ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken cancellationToken)
+            {
+                if (_firstByteStatus == FirstByteStatus.Available)
+                {
+                    if (buffer.Length != 0)
+                    {
+                        buffer.Span[0] = _firstByte;
+                        _firstByteStatus = FirstByteStatus.Consumed;
+                        return new ValueTask<int>(1);
+                    }
+
+                    return new ValueTask<int>(0);
+                }
+
+                Debug.Assert(_firstByteStatus == FirstByteStatus.Consumed);
+                return _stream.ReadAsync(buffer, cancellationToken);
+            }
+
+            public override async Task CopyToAsync(Stream destination, int bufferSize, CancellationToken cancellationToken)
+            {
+                Debug.Assert(_firstByteStatus != FirstByteStatus.None);
```

Comment:
does the code that uses the stream from GetDecompressionStream() need all these members implemented? eg it is passed to arbitrary user code?

Reply:
to answer my question, I it is only used through ZLibOrDeflateStream, which only calls PeekFirstByteXX(), Seek, SetLength, and Flush on PeekFirstByteReadStream. The rest can be debug.fail.